File: permutations.cpp
#include "leetcode.h"
class Solution {
public:
    vector<vector<int> > permute(vector<int> &num) {
        vector<vector<int> > result;
        permute(num, 0, result);
        return result;
    }
    void permute(vector<int> &num, int index,
            vector<vector<int> > &result) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (num.size() == index) {
            result.push_back(num);
            return;
        }
        for (int i=index; i < num.size(); ++i) {
            int temp = num[i];
            num[i] = num[index];
            num[index] = temp;
            permute(num, index+1, result);
            temp = num[i];
            num[i] = num[index];
            num[index] = temp;
        }
    }
    //vector<vector<int> > permute(vector<int> &num) {
    //    // Start typing your C/C++ solution below
    //    // DO NOT write int main() function
    //    vector<vector<int> > result;
    //    vector<int> temp;
    //    if (num.size() == 1) {
    //        temp.push_back(num[0]);
    //        result.push_back(temp);
    //        return result;
    //    }
    //    vector<int> sub(num.begin()+1, num.end());
    //    vector<vector<int> > v = permute(sub);
    //    for (int i=0; i < v.size(); ++i) {
    //        for (int j=0; j < v[i].size(); j++) {
    //            temp.clear();
    //            temp = v[i];
    //            int tmp = temp[j];
    //            temp[j] = num[0];
    //            temp.insert(temp.begin(), tmp);
    //            result.push_back(temp);
    //        }
    //        temp = v[i];
    //        temp.insert(temp.begin(), num[0]);
    //        result.push_back(temp);
    //    }
    //    return result;
    //}
    //vector<vector<int> > permute(vector<int> &num) {
    //    // Start typing your C/C++ solution below
    //    // DO NOT write int main() function
    //    vector<vector<int> > result;
    //    vector<int> temp;
    //    if (num.size() == 1) {
    //        temp.push_back(num[0]);
    //        result.push_back(temp);
    //        return result;
    //    }
    //    vector<int> sub(num.begin()+1, num.end());
    //    vector<vector<int> > v = permute(sub);
    //    for (int i=0; i < v.size(); ++i) {
    //        for (int j=0; j <= v[i].size(); j++) {
    //            temp = v[i];
    //            temp.insert(temp.begin()+j, num[0]);
    //            result.push_back(temp);
    //        }
    //    }
    //    return result;
    //}
};
int main() {
    Solution s;
    vector<int> in;
    in.push_back(1);
    in.push_back(2);
    in.push_back(3);
    vector<vector<int> > v;
    v = s.permute(in);
    for (int i=0; i < v.size(); ++i) {
        for (int j=0; j < v[i].size(); j++) {
            cout << v[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}

File: permutations.java
import java.util.*;
public class Solution {
    public ArrayList<ArrayList<Integer>> permute(int[] num) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        permute(num, 0, result);
        return result;
    }
    public void permute(int[] num, int index, ArrayList<ArrayList<Integer>> result) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if (num.length == index) {
            ArrayList<Integer> temp = new ArrayList<Integer>();
            for (int i = 0; i < num.length; i++) {
                temp.add(num[i]);
            }
            result.add(temp);
            return;
        }
        for (int i=index; i < num.length; ++i) {
            int tmp = num[i];
            num[i] = num[index];
            num[index] = tmp;
            permute(num, index+1, result);
            tmp = num[i];
            num[i] = num[index];
            num[index] = tmp;
        }
    }
    //public ArrayList<ArrayList<Integer>> permute(int[] num) {
    //    // Start typing your Java solution below
    //    // DO NOT write main() function
    //    ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
    //    ArrayList<Integer> temp = new ArrayList<Integer>();
    //    if (num.length == 1) {
    //        temp.add(num[0]);
    //        result.add(temp);
    //        return result;
    //    }
    //    int[] sub = new int[num.length-1];
    //    for (int i=0; i < sub.length; i++) {
    //        sub[i] = num[i+1];
    //    }
    //    ArrayList<ArrayList<Integer>> v = permute(sub);
    //    for (int i=0; i < v.size(); ++i) {
    //        for (int j=0; j < v.get(i).size(); j++) {
    //            temp = (ArrayList<Integer>)v.get(i).clone();
    //            int tmp = temp.get(j);
    //            temp.set(j, num[0]);
    //            temp.add(0, tmp);
    //            result.add(temp);
    //        }
    //        temp = (ArrayList<Integer>)v.get(i).clone();
    //        temp.add(0, num[0]);
    //        result.add(temp);
    //    }
    //    return result;
    //}
    //public ArrayList<ArrayList<Integer>> permute(int[] num) {
    //    // Start typing your Java solution below
    //    // DO NOT write main() function
    //    ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
    //    ArrayList<Integer> temp = new ArrayList<Integer>();
    //    if (num.length == 1) {
    //        temp.add(num[0]);
    //        result.add(temp);
    //        return result;
    //    }
    //    int[] sub = new int[num.length-1];
    //    for (int i=0; i < sub.length; i++) {
    //        sub[i] = num[i+1];
    //    }
    //    ArrayList<ArrayList<Integer>> v = permute(sub);
    //    for (int i=0; i < v.size(); ++i) {
    //        for (int j=0; j <= v.get(i).size(); j++) {
    //            temp = (ArrayList<Integer>)v.get(i).clone();
    //            temp.add(j, num[0]);
    //            result.add(temp);
    //        }
    //    }
    //    return result;
    //}
}
